packages/ra-core/src/dataProvider/useMutation.ts

```diff
@@ -129,11 +178,131 @@ const useMutation = (
+        loading: boolean;
+        loaded: boolean;
+    }
+];
+
+/**
+ * Utility function for merging parameters
+ *
+ * useMutation() parameters can be passed:
+ * - at definition time (e.g. useMutation({ type: 'update', resource: 'posts', payload: { id: 1, data: { title: '' } } }) )
+ * - at call time (e.g [mutate] = useMutation(); mutate({ type: 'update', resource: 'posts', payload: { id: 1, data: { title: '' } } }))
+ * - both
+ *
+ * This function merges the definition time and call time parameters.
+ *
+ * This is useful because useMutation() is used by higher-level hooks like
+ * useCreate() or useUpade(), and these hooks can be called both ways.
+ * So it makes sense to make useMutation() capable of handling both call types
+ * as it avoids repetition higher in the hook chain.
+ *
+ * Also, the call time query may be a DOM Event if the callback is used
+ * as an event listener, as in:
+ *
+ * const UpdateButton = () => {
+ *     const mutate = useMutation({ type: 'update', resource: 'posts', payload: { id: 1, data: { title: '' } } });
+ *     return <button onclick={mutate}>Click me</button>
+ * };
+ *
+ * This usage is accepted, and therefore this function checks if the call time
+ * query is an Event, and discards it in that case.
+ *
+ * @param query {Mutation}
+ * @param callTimeQuery {Mutation}
+ * @param options {Object}
+ * @param callTimeOptions {Object}
+ *
+ * @return { type, resource, payload, options } The merged parameters
+ */
+const mergeDefinitionAndCallTimeParameters = (
+    query?: Mutation,
+    callTimeQuery?: Mutation | Event,
+    options?: MutationOptions,
+    callTimeOptions?: MutationOptions
+) => {
+    if (!query && (!callTimeQuery || callTimeQuery instanceof Event)) {
+        throw new Error('Missing query either at definition or at call time');
+    }
+    if (callTimeQuery instanceof Event)
+        return {
+            type: query.type,
+            resource: query.resource,
+            payload: query.payload,
+            options: sanitizeOptions(options),
+        };
+    if (query)
+        return {
+            type: query.type || callTimeQuery.type,
+            resource: query.resource || callTimeQuery.resource,
+            payload: callTimeQuery
+                ? merge({}, query.payload, callTimeQuery.payload)
+                : query.payload,
+            options: callTimeOptions
+                ? merge(
+                      {},
+                      sanitizeOptions(options),
+                      sanitizeOptions(callTimeOptions)
+                  )
+                : sanitizeOptions(options),
+        };
+    return {
+        type: callTimeQuery.type,
+        resource: callTimeQuery.resource,
+        payload: callTimeQuery.payload,
+        options: sanitizeOptions(callTimeOptions),
+    };
+};
+
+const hasDeclarativeSideEffectsSupport = (
+    options?: MutationOptions,
+    callTimeOptions?: MutationOptions
+) => {
+    if (!options && !callTimeOptions) return false;
```

Comment:
I don't think so. Can you elaborate?

Reply:
when `(!options && !callTimeOptions)` is true, none of the two following tests can be true (since the need at least `callTimeOptions`  or `options` to exist)

Therefore, the last `return false` is sufficient. 
In other words : by default `false` is returned, except when `(callTimeOptions && callTimeOptions.withDeclarativeSideEffectsSupport)` or `(options && options.withDeclarativeSideEffectsSupport)`